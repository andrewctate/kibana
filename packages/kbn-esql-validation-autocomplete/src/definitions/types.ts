/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import type { ESQLCommand, ESQLCommandOption, ESQLFunction, ESQLMessage } from '@kbn/esql-ast';

export type FunctionParameterType =
  | 'number'
  | 'date'
  | 'string'
  | 'boolean'
  | 'null'
  | 'any'
  | 'ip'
  | 'chrono_literal'
  | 'time_literal'
  | 'cartesian_point'
  | 'geo_point'
  | 'number[]'
  | 'string[]'
  | 'boolean[]'
  | 'date[]';

export interface FunctionDefinition {
  type: 'builtin' | 'agg' | 'eval';
  ignoreAsSuggestion?: boolean;
  name: string;
  alias?: string[];
  description: string;
  supportedCommands: string[];
  supportedOptions?: string[];
  signatures: Array<{
    params: Array<{
      name: string;
      type: FunctionParameterType;
      optional?: boolean;
      noNestingFunctions?: boolean;
      supportsWildcard?: boolean;
      literalOnly?: boolean;
    }>;
    minParams?: number;
    returnType: string;
    examples?: string[];
  }>;
  validate?: (fnDef: ESQLFunction) => ESQLMessage[];
}

type CommandParameterType =
  | 'any'
  | 'source'
  | 'function'
  | 'column'
  | 'number'
  | 'string'
  | 'boolean';

export interface CommandBaseDefinition {
  name: string;
  alias?: string;
  description: string;
  signature: {
    multipleParams: boolean;
    // innerType here is useful to drill down the type in case of "column"
    // i.e. column of type string
    params: Array<{
      name: string;
      type: CommandParameterType;
      optional?: boolean;
      innerType?: string;
      values?: string[];
      valueDescriptions?: string[];
      literalOnly?: boolean;
      wildcards?: boolean;
    }>;
  };
}

export interface CommandOptionsDefinition extends CommandBaseDefinition {
  wrapped?: string[];
  optional: boolean;
  skipCommonValidation?: boolean;
  validate?: (
    option: ESQLCommandOption,
    command: ESQLCommand,
    references?: unknown
  ) => ESQLMessage[];
}

export interface CommandModeDefinition {
  name: string;
  description: string;
  values: Array<{ name: string; description: string }>;
  prefix?: string;
}

export interface CommandDefinition extends CommandBaseDefinition {
  options: CommandOptionsDefinition[];
  examples: string[];
  validate?: (option: ESQLCommand) => ESQLMessage[];
  modes: CommandModeDefinition[];
}

export interface Literals {
  name: string;
  description: string;
}

// TODO - think about separate types for functions and commands
export type SignatureType =
  | FunctionDefinition['signatures'][number]
  | CommandOptionsDefinition['signature'];
export type SignatureArgType = SignatureType['params'][number];
