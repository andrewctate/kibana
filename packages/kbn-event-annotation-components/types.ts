/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { $Values } from 'utility-types';
import { DataViewSpec, KibanaQueryOutput } from '@kbn/data-plugin/common';
import { UserContentCommonSchema } from '@kbn/content-management-table-list-view-table';
import { LineStyle } from '@kbn/visualization-ui-components';
import { SavedObjectsFindOptionsReference } from '@kbn/core-saved-objects-api-server';
import { ExpressionAstExpression } from '@kbn/expressions-plugin/common';
import { SavedObjectCommon } from '@kbn/saved-objects-finder-plugin/common';
import { AvailableAnnotationIcons } from './constants';

export type AvailableAnnotationIcon = $Values<typeof AvailableAnnotationIcons>;

export type PointStyleProps = StyleSharedProps & {
  icon?: AvailableAnnotationIcon;
  lineWidth?: number;
  lineStyle?: LineStyle;
  textVisibility?: boolean;
};

export type PointInTimeEventAnnotationConfig = {
  id: string;
  type: ManualAnnotationType;
  key: {
    type: 'point_in_time';
    timestamp: string;
  };
} & PointStyleProps;

export type Fill = 'inside' | 'outside' | 'none';
export type ManualAnnotationType = 'manual';
export type QueryAnnotationType = 'query';
export type KeyType = 'point_in_time' | 'range';

interface StyleSharedProps {
  label: string;
  color?: string;
  isHidden?: boolean;
}

export type RangeStyleProps = StyleSharedProps & {
  outside?: boolean;
};

export type RangeEventAnnotationConfig = {
  type: ManualAnnotationType;
  id: string;
  key: {
    type: 'range';
    timestamp: string;
    endTimestamp: string;
  };
} & RangeStyleProps;

export type StyleProps = PointStyleProps & RangeStyleProps;

export type QueryPointEventAnnotationConfig = {
  id: string;
  type: QueryAnnotationType;
  filter: KibanaQueryOutput;
  timeField?: string;
  textField?: string;
  extraFields?: string[];
  key: {
    type: 'point_in_time';
  };
} & PointStyleProps;

export type EventAnnotationConfig =
  | PointInTimeEventAnnotationConfig
  | RangeEventAnnotationConfig
  | QueryPointEventAnnotationConfig;

export interface EventAnnotationGroupConfig {
  annotations: EventAnnotationConfig[];
  indexPatternId: string;
  ignoreGlobalFilters: boolean;
  title: string;
  description: string;
  tags: string[];
  dataViewSpec?: DataViewSpec;
}

export type EventAnnotationGroupContent = UserContentCommonSchema & {
  attributes: { indexPatternId: string; dataViewSpec?: DataViewSpec };
};

export interface EventAnnotationServiceType {
  loadAnnotationGroup: (savedObjectId: string) => Promise<EventAnnotationGroupConfig>;
  groupExistsWithTitle: (title: string) => Promise<boolean>;
  findAnnotationGroupContent: (
    searchTerm: string,
    pageSize: number,
    references?: SavedObjectsFindOptionsReference[],
    referencesToExclude?: SavedObjectsFindOptionsReference[]
  ) => Promise<{ total: number; hits: EventAnnotationGroupContent[] }>;
  deleteAnnotationGroups: (ids: string[]) => Promise<void>;
  createAnnotationGroup: (group: EventAnnotationGroupConfig) => Promise<{ id: string }>;
  updateAnnotationGroup: (
    group: EventAnnotationGroupConfig,
    savedObjectId: string
  ) => Promise<void>;
  toExpression: (props: EventAnnotationConfig[]) => ExpressionAstExpression[];
  toFetchExpression: (props: {
    interval: string;
    groups: Array<
      Pick<EventAnnotationGroupConfig, 'annotations' | 'ignoreGlobalFilters' | 'indexPatternId'>
    >;
  }) => ExpressionAstExpression[];
  renderEventAnnotationGroupSavedObjectFinder: (props: {
    fixedPageSize?: number;
    onChoose: (value: {
      id: string;
      type: string;
      fullName: string;
      savedObject: SavedObjectCommon<unknown>;
    }) => void;
    onCreateNew: () => void;
  }) => JSX.Element;
}
