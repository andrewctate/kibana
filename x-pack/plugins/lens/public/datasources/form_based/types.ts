/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
import type { DragDropIdentifier } from '@kbn/dom-drag-drop';
import type { GenericIndexPatternColumn } from './operations';
import type { IndexPattern, IndexPatternField, DragDropOperation } from '../../types';

export type {
  GenericIndexPatternColumn,
  OperationType,
  IncompleteColumn,
  FieldBasedIndexPatternColumn,
  FiltersIndexPatternColumn,
  RangeIndexPatternColumn,
  TermsIndexPatternColumn,
  DateHistogramIndexPatternColumn,
  MinIndexPatternColumn,
  MaxIndexPatternColumn,
  AvgIndexPatternColumn,
  CardinalityIndexPatternColumn,
  SumIndexPatternColumn,
  MedianIndexPatternColumn,
  StandardDeviationIndexPatternColumn,
  PercentileIndexPatternColumn,
  PercentileRanksIndexPatternColumn,
  CountIndexPatternColumn,
  LastValueIndexPatternColumn,
  CumulativeSumIndexPatternColumn,
  CounterRateIndexPatternColumn,
  DerivativeIndexPatternColumn,
  MovingAverageIndexPatternColumn,
  FormulaIndexPatternColumn,
  MathIndexPatternColumn,
  OverallSumIndexPatternColumn,
  StaticValueIndexPatternColumn,
  TimeScaleIndexPatternColumn,
} from './operations';

export type { FormulaPublicApi } from './operations/definitions/formula/formula_public_api';

export type DraggedField = DragDropIdentifier & {
  field: IndexPatternField;
  indexPatternId: string;
};

export type PersistedIndexPatternLayer = Omit<FormBasedLayer, 'indexPatternId'>;

export interface FormBasedPrivateState {
  currentIndexPatternId: string;
  layers: Record<string, FormBasedLayer>;
}

export interface DataViewDragDropOperation extends DragDropOperation {
  dataView: IndexPattern;
  column?: GenericIndexPatternColumn;
}
