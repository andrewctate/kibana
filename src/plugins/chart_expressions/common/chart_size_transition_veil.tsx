/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { useCallback, useRef, useState } from 'react';
import fastIsEqual from 'fast-deep-equal';
import type { ChartSizeSpec } from './types';

/**
 * This hook is used to show a veil over the chart while it is being resized
 * in response to a change in the container dimensions.
 *
 * It is only relevant if client dimensions are being requested based on chart configuration.
 */
export function useSizeTransitionVeil(
  chartSizeSpec: ChartSizeSpec,
  setChartSize: (d: ChartSizeSpec) => void
) {
  const [showVeil, setShowVeil] = useState(false);
  const currentChartSizeSpec = useRef<ChartSizeSpec>();

  if (!fastIsEqual(chartSizeSpec, currentChartSizeSpec.current)) {
    // If the dimensions have changed we request new dimensions from the client
    // and set off a chain of events:
    //
    // 1. we show the veil to hide step 4
    // 2. the charts library will plan a render
    // 3. the client will resize the container
    // 4. the charts library will render the chart based on the original container dimensions
    // 5. the charts library will resize the chart to the updated container dimensions
    // 6. we hide the veil

    setChartSize(chartSizeSpec);
    setShowVeil(true);
    currentChartSizeSpec.current = chartSizeSpec;
  }

  const onResize = useCallback(() => {
    setShowVeil(false);
  }, []);

  return {
    veil: (
      <div
        css={{
          height: '100%',
          width: '100%',
          backgroundColor: 'white',
          position: 'absolute',
          zIndex: 1,
          display: showVeil ? 'block' : 'none',
        }}
      />
    ),
    onResize,
  };
}
