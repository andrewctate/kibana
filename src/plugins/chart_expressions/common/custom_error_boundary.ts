/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';

interface Props {
  onError: (error: Error) => void;
}

export class CustomErrorBoundary extends React.Component<Props> {
  onError: Props['onError'];

  constructor(props: Props) {
    super(props);
    this.onError = props.onError;
  }

  componentDidCatch(error: Error) {
    this.onError(error);
  }

  render() {
    return this.props.children;
  }
}
