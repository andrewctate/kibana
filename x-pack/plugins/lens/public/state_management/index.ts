/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { configureStore, getDefaultMiddleware, Middleware, PreloadedState } from '@reduxjs/toolkit';
import { createLogger } from 'redux-logger';
import { useDispatch, useSelector, TypedUseSelectorHook } from 'react-redux';
import { makeLensReducer, lensActions } from './lens_slice';
import { timeRangeMiddleware } from './time_range_middleware';
import { optimizingMiddleware } from './optimizing_middleware';
import { LensState, LensStoreDeps } from './types';
import { initMiddleware } from './init_middleware';
import { DiffLoggerWindow } from './diff_logger';
export * from './types';
export * from './selectors';

export const {
  loadInitial,
  navigateAway,
  setState,
  setSaveable,
  onActiveDataChange,
  updateState,
  updateDatasourceState,
  updateVisualizationState,
  insertLayer,
  switchVisualization,
  rollbackSuggestion,
  submitSuggestion,
  switchDatasource,
  setToggleFullscreen,
  initEmpty,
  editVisualizationAction,
  removeLayers,
  removeOrClearLayer,
  addLayer,
  setLayerDefaultDimension,
} = lensActions;

let loggerWindow: DiffLoggerWindow;

const diffLogger: Middleware = (store) => (next) => (action) => {
  const prevState = store.getState();
  const result = next(action);
  if (
    (window as unknown as Window & { ELASTIC_LENS_DIFF_LOGGER: boolean }).ELASTIC_LENS_DIFF_LOGGER
  ) {
    if (!loggerWindow) {
      loggerWindow = new DiffLoggerWindow();
    }

    loggerWindow.printDiff({
      actionType: action.type,
      prev: prevState,
      next: store.getState(),
    });
  }
  return result;
};

export const makeConfigureStore = (
  storeDeps: LensStoreDeps,
  preloadedState: PreloadedState<LensState>
) => {
  const middleware = [
    ...getDefaultMiddleware({
      serializableCheck: false,
    }),
    initMiddleware(storeDeps),
    optimizingMiddleware(),
    timeRangeMiddleware(storeDeps.lensServices.data),
  ];
  if (process.env.NODE_ENV === 'development') {
    middleware.push(
      createLogger({
        // @ts-ignore
        predicate: () => window.ELASTIC_LENS_LOGGER,
      })
    );
    middleware.push(diffLogger);
  }

  return configureStore({
    reducer: {
      lens: makeLensReducer(storeDeps),
    },
    middleware,
    preloadedState,
  });
};

type Awaited<T> = T extends PromiseLike<infer U> ? U : T; // TODO - use built-in when we get to TS 4.5
export type LensRootStore = Awaited<ReturnType<typeof makeConfigureStore>>;

export type LensDispatch = LensRootStore['dispatch'];
export type LensGetState = LensRootStore['getState'];
export type LensRootState = ReturnType<LensGetState>;

export const useLensDispatch = () => useDispatch<LensDispatch>();
export const useLensSelector: TypedUseSelectorHook<LensRootState> = useSelector;
