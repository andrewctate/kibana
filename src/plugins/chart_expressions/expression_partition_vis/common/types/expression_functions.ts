/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import {
  ExpressionFunctionDefinition,
  Datatable,
  ExpressionValueRender,
  ExpressionValueBoxed,
} from '@kbn/expressions-plugin/common';
import { ExpressionValueVisDimension } from '@kbn/visualizations-plugin/common';
import {
  PARTITION_LABELS_VALUE,
  PIE_VIS_EXPRESSION_NAME,
  TREEMAP_VIS_EXPRESSION_NAME,
  MOSAIC_VIS_EXPRESSION_NAME,
  WAFFLE_VIS_EXPRESSION_NAME,
  PARTITION_LAYER_BUCKETS_NAME,
} from '../constants';
import {
  RenderValue,
  PieVisConfig,
  LabelPositions,
  ValueFormats,
  TreemapVisConfig,
  MosaicVisConfig,
  WaffleVisConfig,
} from './expression_renderers';

export interface PartitionLabelsArguments {
  show: boolean;
  position: LabelPositions;
  values: boolean;
  valuesFormat: ValueFormats;
  percentDecimals: number;
  /** @deprecated This field is deprecated and going to be removed in the futher release versions. */
  truncate?: number | null;
  /** @deprecated This field is deprecated and going to be removed in the futher release versions. */
  last_level?: boolean;
}

export type ExpressionValuePartitionLabels = ExpressionValueBoxed<
  typeof PARTITION_LABELS_VALUE,
  {
    show: boolean;
    position: LabelPositions;
    values: boolean;
    valuesFormat: ValueFormats;
    percentDecimals: number;
    /** @deprecated This field is deprecated and going to be removed in the futher release versions. */
    truncate?: number | null;
    /** @deprecated This field is deprecated and going to be removed in the futher release versions. */
    last_level?: boolean;
  }
>;

export interface PartitionLayerBucketArguments {
  metric: ExpressionValueVisDimension | string;
  bucket: ExpressionValueVisDimension | string; // TODO - not sure why the buckets column was considered optional, but preserving for now
}

export type ExpressionValuePartitionLayerBucket = ExpressionValueBoxed<
  'partition_layer_bucket',
  PartitionLayerBucketArguments
>;

export type PartitionLayerBucketExpressionFunctionDefinition = ExpressionFunctionDefinition<
  typeof PARTITION_LAYER_BUCKETS_NAME,
  null,
  PartitionLayerBucketArguments,
  ExpressionValuePartitionLayerBucket
>;

export interface PartitionLayerFieldsArguments {
  fields: Array<ExpressionValueVisDimension | string>;
}

export type ExpressionValuePartitionLayerFields = ExpressionValueBoxed<
  'partition_layer_fields',
  PartitionLayerFieldsArguments
>;

export type PartitionLayerFieldsExpressionFunctionDefinition = ExpressionFunctionDefinition<
  typeof PARTITION_LAYER_BUCKETS_NAME,
  null,
  PartitionLayerFieldsArguments,
  ExpressionValuePartitionLayerFields
>;

export type PieVisExpressionFunctionDefinition = ExpressionFunctionDefinition<
  typeof PIE_VIS_EXPRESSION_NAME,
  Datatable,
  PieVisConfig,
  ExpressionValueRender<RenderValue>
>;

export type TreemapVisExpressionFunctionDefinition = ExpressionFunctionDefinition<
  typeof TREEMAP_VIS_EXPRESSION_NAME,
  Datatable,
  TreemapVisConfig,
  ExpressionValueRender<RenderValue>
>;

export type MosaicVisExpressionFunctionDefinition = ExpressionFunctionDefinition<
  typeof MOSAIC_VIS_EXPRESSION_NAME,
  Datatable,
  MosaicVisConfig,
  ExpressionValueRender<RenderValue>
>;

export type WaffleVisExpressionFunctionDefinition = ExpressionFunctionDefinition<
  typeof WAFFLE_VIS_EXPRESSION_NAME,
  Datatable,
  WaffleVisConfig,
  ExpressionValueRender<RenderValue>
>;

export enum ChartTypes {
  PIE = 'pie',
  DONUT = 'donut',
  TREEMAP = 'treemap',
  MOSAIC = 'mosaic',
  WAFFLE = 'waffle',
}
