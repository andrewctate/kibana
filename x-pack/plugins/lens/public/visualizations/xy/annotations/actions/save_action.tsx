/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React, { useState } from 'react';
import { i18n } from '@kbn/i18n';
import { render, unmountComponentAtNode } from 'react-dom';
import { EventAnnotationServiceType } from '@kbn/event-annotation-plugin/public';
import { ToastsStart } from '@kbn/core-notifications-browser';
import { MountPoint } from '@kbn/core-mount-utils-browser';
import { FormattedMessage } from '@kbn/i18n-react';
import { SavedObjectSaveModal } from '@kbn/saved-objects-plugin/public';
import { EVENT_ANNOTATION_GROUP_TYPE } from '@kbn/event-annotation-plugin/common';
import { EuiIcon } from '@elastic/eui';
import { type SavedObjectTaggingPluginStart } from '@kbn/saved-objects-tagging-plugin/public';
import type { LayerAction, StateSetter } from '../../../../types';
import { XYByReferenceAnnotationLayerConfig, XYAnnotationLayerConfig, XYState } from '../../types';

// exported for testing only
export const SaveModal = ({
  domElement,
  state,
  layer,
  setState,
  eventAnnotationService,
  toasts,
  savedObjectsTagging,
}: {
  domElement: HTMLDivElement;
  state: XYState;
  layer: XYAnnotationLayerConfig;
  setState: StateSetter<XYState, unknown>;
  eventAnnotationService: EventAnnotationServiceType;
  toasts: ToastsStart;
  savedObjectsTagging?: SavedObjectTaggingPluginStart;
}) => {
  const initialTags: string[] = [];

  const [selectedTags, setSelectedTags] = useState<string[]>(initialTags);

  const closeModal = () => unmountComponentAtNode(domElement);

  return (
    <SavedObjectSaveModal
      onSave={async ({ newTitle, newDescription }) => {
        let savedId: string;

        try {
          const { id } = await eventAnnotationService.createAnnotationGroup({
            ...layer,
            title: newTitle,
            description: newDescription,
            tags: selectedTags,
          });

          savedId = id;
        } catch (err) {
          toasts.addError(err, {
            title: i18n.translate(
              'xpack.lens.xyChart.annotations.saveAnnotationGroupToLibrary.errorToastTitle',
              {
                defaultMessage: 'Failed to save "{title}"',
                values: {
                  title: newTitle,
                },
              }
            ),
          });

          return;
        }

        const newLayer: XYByReferenceAnnotationLayerConfig = {
          ...layer,
          annotationGroupId: savedId,
          __lastSaved: {
            ...layer,
            title: newTitle,
            description: newDescription,
            tags: selectedTags,
          },
        };

        setState({
          ...state,
          layers: state.layers.map((existingLayer) =>
            existingLayer.layerId === newLayer.layerId ? newLayer : existingLayer
          ),
        });

        closeModal();

        toasts.addSuccess({
          title: i18n.translate(
            'xpack.lens.xyChart.annotations.saveAnnotationGroupToLibrary.successToastTitle',
            {
              defaultMessage: 'Saved "{title}"',
              values: {
                title: newTitle,
              },
            }
          ),
          text: ((element) =>
            render(
              <div>
                <FormattedMessage
                  id="xpack.lens.xyChart.annotations.saveAnnotationGroupToLibrary.successToastBody"
                  defaultMessage="View or manage in the {link}"
                  values={{
                    link: <a href="#">annotation library</a>,
                  }}
                />
              </div>,
              element
            )) as MountPoint,
        });
      }}
      onClose={closeModal}
      title=""
      showCopyOnSave={false}
      objectType={EVENT_ANNOTATION_GROUP_TYPE}
      showDescription={true}
      confirmButtonLabel={
        <>
          <div>
            <EuiIcon type="save" />
          </div>
          <div>
            {i18n.translate(
              'xpack.lens.xyChart.annotations.saveAnnotationGroupToLibrary.confirmButton',
              { defaultMessage: 'Save annotation group' }
            )}
          </div>
        </>
      }
      options={
        savedObjectsTagging ? (
          <savedObjectsTagging.ui.components.SavedObjectSaveModalTagSelector
            initialSelection={initialTags}
            onTagsSelected={setSelectedTags}
          />
        ) : undefined
      }
    />
  );
};

export const getSaveLayerAction = (props: {
  state: XYState;
  layer: XYAnnotationLayerConfig;
  setState: StateSetter<XYState, unknown>;
  eventAnnotationService: EventAnnotationServiceType;
  isNew?: boolean;
  toasts: ToastsStart;
  savedObjectsTagging?: SavedObjectTaggingPluginStart;
}): LayerAction => {
  const displayName = props.isNew
    ? i18n.translate('xpack.lens.xyChart.annotations.addAnnotationGroupToLibrary', {
        defaultMessage: 'Add to library',
      })
    : i18n.translate('xpack.lens.xyChart.annotations.saveAnnotationGroupToLibrary', {
        defaultMessage: 'Save to library',
      });
  return {
    displayName,
    description: i18n.translate(
      'xpack.lens.xyChart.annotations.addAnnotationGroupToLibraryDescription',
      { defaultMessage: 'Saves annotation group as separate saved object' }
    ),
    execute: async (domElement) => {
      if (props.isNew && domElement) {
        render(<SaveModal {...props} domElement={domElement} />, domElement);
      } else {
        return props.eventAnnotationService.createAnnotationGroup(props.layer).then();
      }
    },
    icon: 'save',
    isCompatible: true,
    'data-test-subj': 'lnsXY_annotationLayer_saveToLibrary',
  };
};
