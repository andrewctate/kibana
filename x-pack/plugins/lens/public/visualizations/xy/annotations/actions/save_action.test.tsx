/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React from 'react';
import { toastsServiceMock } from '@kbn/core-notifications-browser-mocks/src/toasts_service.mock';
import { EventAnnotationServiceType } from '@kbn/event-annotation-plugin/public';
import { XYByValueAnnotationLayerConfig, XYAnnotationLayerConfig, XYState } from '../../types';
import { SaveModal } from './save_action';
import { shallowWithIntl } from '@kbn/test-jest-helpers';
import { PointInTimeEventAnnotationConfig } from '@kbn/event-annotation-plugin/common';
import { SavedObjectSaveModal } from '@kbn/saved-objects-plugin/public';
import { taggingApiMock } from '@kbn/saved-objects-tagging-plugin/public/mocks';

describe('annotation group save action', () => {
  const layerId = 'mylayerid';

  const layer: XYByValueAnnotationLayerConfig = {
    layerId,
    layerType: 'annotations',
    indexPatternId: 'some-index-pattern',
    ignoreGlobalFilters: false,
    annotations: [
      {
        id: 'some-annotation-id',
        type: 'manual',
        key: {
          type: 'point_in_time',
          timestamp: 'timestamp',
        },
      } as PointInTimeEventAnnotationConfig,
    ],
  };

  const savedId = 'saved-id-123';

  const getProps: () => Parameters<typeof SaveModal>[0] = () => ({
    state: {
      preferredSeriesType: 'area',
      legend: { isVisible: true, position: 'bottom' },
      layers: [{ layerId } as XYAnnotationLayerConfig],
    } as XYState,
    layer,
    setState: jest.fn(),
    eventAnnotationService: {
      createAnnotationGroup: jest.fn(() => Promise.resolve({ id: savedId })),
      loadAnnotationGroup: jest.fn(),
      toExpression: jest.fn(),
      toFetchExpression: jest.fn(),
      renderEventAnnotationGroupSavedObjectFinder: jest.fn(),
    } as EventAnnotationServiceType,
    toasts: toastsServiceMock.createStartContract(),
    domElement: document.createElement('div'),
    savedObjectsTagging: taggingApiMock.create(),
  });

  let props: ReturnType<typeof getProps>;
  beforeEach(() => {
    props = getProps();
  });

  const modalSaveArgs = {
    newCopyOnSave: false,
    isTitleDuplicateConfirmed: false,
    onTitleDuplicate: () => {},
  };

  test('successful initial save', async () => {
    const wrapper = shallowWithIntl(<SaveModal {...props} />);

    const newTitle = 'title';
    const newDescription = 'description';
    const myTags = ['my', 'many', 'tags'];

    (wrapper
      .find(SavedObjectSaveModal)
      .prop('options') as React.ReactElement)!.props.onTagsSelected(myTags);

    // ignore the linter, you need this await statement
    await wrapper.find(SavedObjectSaveModal).prop('onSave')({
      newTitle,
      newDescription,
      ...modalSaveArgs,
    });

    expect(props.eventAnnotationService.createAnnotationGroup).toHaveBeenCalledWith({
      ...props.layer,
      title: newTitle,
      description: newDescription,
      tags: myTags,
    });

    expect((props.setState as jest.Mock).mock.calls).toMatchSnapshot();

    expect(props.toasts.addSuccess).toHaveBeenCalledTimes(1);
  });

  test('failed initial save', async () => {
    (props.eventAnnotationService.createAnnotationGroup as jest.Mock).mockRejectedValue(
      new Error('oh noooooo')
    );

    const wrapper = shallowWithIntl(<SaveModal {...props} />);

    const newTitle = 'title';
    const newDescription = 'new description';

    // ignore the linter, you need this await statement
    await wrapper.find(SavedObjectSaveModal).prop('onSave')({
      newTitle,
      newDescription,
      ...modalSaveArgs,
    });

    expect(props.eventAnnotationService.createAnnotationGroup).toHaveBeenCalledWith({
      ...props.layer,
      title: newTitle,
      description: newDescription,
      tags: [],
    });

    expect(props.toasts.addError).toHaveBeenCalledTimes(1);

    expect(props.setState).not.toHaveBeenCalled();

    expect(props.toasts.addSuccess).not.toHaveBeenCalled();
  });
});
