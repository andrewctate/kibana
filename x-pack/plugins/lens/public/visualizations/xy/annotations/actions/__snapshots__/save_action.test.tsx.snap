// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`annotation group save action successful initial save 1`] = `
Array [
  Array [
    Object {
      "layers": Array [
        Object {
          "__lastSaved": Object {
            "annotations": Array [
              Object {
                "id": "some-annotation-id",
                "key": Object {
                  "timestamp": "timestamp",
                  "type": "point_in_time",
                },
                "type": "manual",
              },
            ],
            "description": "description",
            "ignoreGlobalFilters": false,
            "indexPatternId": "some-index-pattern",
            "layerId": "mylayerid",
            "layerType": "annotations",
            "tags": Array [
              "my",
              "many",
              "tags",
            ],
            "title": "title",
          },
          "annotationGroupId": "saved-id-123",
          "annotations": Array [
            Object {
              "id": "some-annotation-id",
              "key": Object {
                "timestamp": "timestamp",
                "type": "point_in_time",
              },
              "type": "manual",
            },
          ],
          "ignoreGlobalFilters": false,
          "indexPatternId": "some-index-pattern",
          "layerId": "mylayerid",
          "layerType": "annotations",
        },
      ],
      "legend": Object {
        "isVisible": true,
        "position": "bottom",
      },
      "preferredSeriesType": "area",
    },
  ],
]
`;
